programs/client/Client.cpp

```diff
@@ -2060,13 +2280,58 @@ class Client : public Poco::Util::Application
 
 }
 
+using signal_function = void(int, siginfo_t*, void*);
+
+/// Setup signal handlers.
+static void add_signal_handler(const std::vector<int> & signals, signal_function handler)
+{
+    struct sigaction sa;
+    memset(&sa, 0, sizeof(sa));
+    sa.sa_sigaction = handler;
+    sa.sa_flags = SA_SIGINFO;
+
+    {
+#if defined(OS_DARWIN)
+        sigemptyset(&sa.sa_mask);
+        for (auto signal : signals)
+            sigaddset(&sa.sa_mask, signal);
+#else
+        if (sigemptyset(&sa.sa_mask))
+            throw Poco::Exception("Cannot set signal handler.");
+
+        for (auto signal : signals)
+            if (sigaddset(&sa.sa_mask, signal))
+                throw Poco::Exception("Cannot set signal handler.");
+#endif
+
+        for (auto signal : signals)
+            if (sigaction(signal, &sa, nullptr))
+                throw Poco::Exception("Cannot set signal handler.");
+    }
+};
+
+/** Handler for "fault" or diagnostic signals. */
+static void signalHandler(int sig, siginfo_t * /*info*/, void * context)
+{
+    const ucontext_t signal_context = *reinterpret_cast<ucontext_t *>(context);
+    const StackTrace stack_trace(signal_context);
+    std::cerr << fmt::format("Received signal {} at: {}", sig,
+        stack_trace.toString()) << std::endl;
```

Comment:
toString is not signal-safe. std::cerr and fmt::format neither.

Reply:
I want to have stack traces when the client dies, but don't want to have a separate thread in client to handle these faults (that's complicated). This is for deadly signals, the app is going to terminate anyway, so maybe it's OK to use not signal-safe functions?

Reply:
Discussed on zoom -- we had that before and it lead to deadlocks. The solution is to copy the handling we have in BaseDaemon.

Reply:
I removed it for now because I want to merge soon and run the fuzzer in CI. I'll add proper signal handling separately.